PartitionServiceSafetyCheckTest.clusterShouldBeEventuallySafe_withPartitionInitialization

Failed on a PR builder against the 4.1.z branch.

http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/4010/testReport/junit/com.hazelcast.internal.partition/PartitionServiceSafetyCheckTest/clusterShouldBeEventuallySafe_withPartitionInitialization/

**Error Message**
```
Instances not in safe state! {[127.0.0.1]:5702=REPLICA_NOT_SYNC}
```
**Stacktrace**
```
java.lang.AssertionError: Instances not in safe state! {[127.0.0.1]:5702=REPLICA_NOT_SYNC}
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at com.hazelcast.test.HazelcastTestSupport.assertAllInSafeState(HazelcastTestSupport.java:769)
	at com.hazelcast.test.HazelcastTestSupport$1.run(HazelcastTestSupport.java:734)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1248)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1265)
	at com.hazelcast.test.HazelcastTestSupport.waitAllForSafeState(HazelcastTestSupport.java:732)
	at com.hazelcast.test.HazelcastTestSupport.waitAllForSafeState(HazelcastTestSupport.java:728)
	at com.hazelcast.internal.partition.PartitionServiceSafetyCheckTest.assertSafeEventually(PartitionServiceSafetyCheckTest.java:209)
	at com.hazelcast.internal.partition.PartitionServiceSafetyCheckTest.clusterShouldBeEventuallySafe_withPartitionInitialization(PartitionServiceSafetyCheckTest.java:114)
	at com.hazelcast.internal.partition.PartitionServiceSafetyCheckTest.clusterShouldBeEventuallySafe_withPartitionInitialization(PartitionServiceSafetyCheckTest.java:90)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```
**Standard Output**
```
Finished Running Test: clusterShouldBeSafe_whenBackupsBlocked_withForceToBeSafe in 6.334 seconds.
12:03:28,273  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [LifecycleService] Thread-141 - [127.0.0.1]:5703 [dev] [4.1.2-SNAPSHOT] [127.0.0.1]:5703 is SHUTTING_DOWN
12:03:28,274  WARN |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5703 [dev] [4.1.2-SNAPSHOT] Terminating forcefully...
12:03:28,274  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5703 [dev] [4.1.2-SNAPSHOT] Shutting down connection manager...
12:03:28,274  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MockServer] Thread-141 - [127.0.0.1]:5702 [dev] [4.1.2-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5703, connection: MockConnection{localEndpoint=[127.0.0.1]:5702, remoteEndpoint=[127.0.0.1]:5703, alive=false}
12:03:28,274  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MockServer] Thread-141 - [127.0.0.1]:5703 [dev] [4.1.2-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5702, connection: MockConnection{localEndpoint=[127.0.0.1]:5703, remoteEndpoint=[127.0.0.1]:5702, alive=false}
12:03:28,274  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MockServer] Thread-141 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5703, connection: MockConnection{localEndpoint=[127.0.0.1]:5704, remoteEndpoint=[127.0.0.1]:5703, alive=false}
12:03:28,274  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MockServer] Thread-141 - [127.0.0.1]:5703 [dev] [4.1.2-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5704, connection: MockConnection{localEndpoint=[127.0.0.1]:5703, remoteEndpoint=[127.0.0.1]:5704, alive=false}
12:03:28,274  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MockServer] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5703, connection: MockConnection{localEndpoint=[127.0.0.1]:5701, remoteEndpoint=[127.0.0.1]:5703, alive=false}
12:03:28,274  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MockServer] Thread-141 - [127.0.0.1]:5703 [dev] [4.1.2-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5701, connection: MockConnection{localEndpoint=[127.0.0.1]:5703, remoteEndpoint=[127.0.0.1]:5701, alive=false}
12:03:28,274  WARN |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MembershipManager] Thread-141 - [127.0.0.1]:5702 [dev] [4.1.2-SNAPSHOT] Member [127.0.0.1]:5703 - ad2db285-ce3b-4aba-8b78-75a5ffb1649e is suspected to be dead for reason: Connection manager is stopped on Member [127.0.0.1]:5703 - ad2db285-ce3b-4aba-8b78-75a5ffb1649e this
12:03:28,274  WARN |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MembershipManager] Thread-141 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Member [127.0.0.1]:5703 - ad2db285-ce3b-4aba-8b78-75a5ffb1649e is suspected to be dead for reason: Connection manager is stopped on Member [127.0.0.1]:5703 - ad2db285-ce3b-4aba-8b78-75a5ffb1649e this
12:03:28,275  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MembershipManager] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Removing Member [127.0.0.1]:5703 - ad2db285-ce3b-4aba-8b78-75a5ffb1649e
12:03:28,276  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [ClusterService] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] 

Members {size:3, ver:5} [
	Member [127.0.0.1]:5701 - ea98e464-aebd-44d0-aaa1-503df99cf13f this
	Member [127.0.0.1]:5704 - a1246091-68ac-407d-951b-8e07dfdb427b
	Member [127.0.0.1]:5702 - c171e63c-d0b4-4794-ac89-682e958bd307
]

12:03:28,276  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [TransactionManagerService] hz.boring_villani.cached.thread-8 - [127.0.0.1]:5702 [dev] [4.1.2-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5703, UUID: ad2db285-ce3b-4aba-8b78-75a5ffb1649e
12:03:28,276  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [TransactionManagerService] hz.relaxed_villani.cached.thread-2 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5703, UUID: ad2db285-ce3b-4aba-8b78-75a5ffb1649e
12:03:28,276  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [ClusterService] hz.relaxed_villani.priority-generic-operation.thread-0 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] 

Members {size:3, ver:5} [
	Member [127.0.0.1]:5701 - ea98e464-aebd-44d0-aaa1-503df99cf13f
	Member [127.0.0.1]:5704 - a1246091-68ac-407d-951b-8e07dfdb427b this
	Member [127.0.0.1]:5702 - c171e63c-d0b4-4794-ac89-682e958bd307
]

12:03:28,276  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [TransactionManagerService] hz.sad_villani.cached.thread-4 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5703, UUID: ad2db285-ce3b-4aba-8b78-75a5ffb1649e
12:03:28,276  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [ClusterService] hz.boring_villani.priority-generic-operation.thread-0 - [127.0.0.1]:5702 [dev] [4.1.2-SNAPSHOT] 

Members {size:3, ver:5} [
	Member [127.0.0.1]:5701 - ea98e464-aebd-44d0-aaa1-503df99cf13f
	Member [127.0.0.1]:5704 - a1246091-68ac-407d-951b-8e07dfdb427b
	Member [127.0.0.1]:5702 - c171e63c-d0b4-4794-ac89-682e958bd307 this
]

12:03:28,276  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5703 [dev] [4.1.2-SNAPSHOT] Shutting down node engine...
12:03:28,285  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [NodeExtension] Thread-141 - [127.0.0.1]:5703 [dev] [4.1.2-SNAPSHOT] Destroying node NodeExtension.
12:03:28,285  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5703 [dev] [4.1.2-SNAPSHOT] Hazelcast Shutdown is completed in 11 ms.
12:03:28,285  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [LifecycleService] Thread-141 - [127.0.0.1]:5703 [dev] [4.1.2-SNAPSHOT] [127.0.0.1]:5703 is SHUTDOWN
12:03:28,285  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [LifecycleService] Thread-141 - [127.0.0.1]:5702 [dev] [4.1.2-SNAPSHOT] [127.0.0.1]:5702 is SHUTTING_DOWN
12:03:28,285  WARN |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5702 [dev] [4.1.2-SNAPSHOT] Terminating forcefully...
12:03:28,285  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5702 [dev] [4.1.2-SNAPSHOT] Shutting down connection manager...
12:03:28,285  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MockServer] Thread-141 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5702, connection: MockConnection{localEndpoint=[127.0.0.1]:5704, remoteEndpoint=[127.0.0.1]:5702, alive=false}
12:03:28,285  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MockServer] Thread-141 - [127.0.0.1]:5702 [dev] [4.1.2-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5704, connection: MockConnection{localEndpoint=[127.0.0.1]:5702, remoteEndpoint=[127.0.0.1]:5704, alive=false}
12:03:28,285  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MockServer] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5702, connection: MockConnection{localEndpoint=[127.0.0.1]:5701, remoteEndpoint=[127.0.0.1]:5702, alive=false}
12:03:28,285  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MockServer] Thread-141 - [127.0.0.1]:5702 [dev] [4.1.2-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5701, connection: MockConnection{localEndpoint=[127.0.0.1]:5702, remoteEndpoint=[127.0.0.1]:5701, alive=false}
12:03:28,285  WARN |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MembershipManager] Thread-141 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Member [127.0.0.1]:5702 - c171e63c-d0b4-4794-ac89-682e958bd307 is suspected to be dead for reason: Connection manager is stopped on Member [127.0.0.1]:5702 - c171e63c-d0b4-4794-ac89-682e958bd307 this
12:03:28,285  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MembershipManager] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Removing Member [127.0.0.1]:5702 - c171e63c-d0b4-4794-ac89-682e958bd307
12:03:28,286  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [ClusterService] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] 

Members {size:2, ver:6} [
	Member [127.0.0.1]:5701 - ea98e464-aebd-44d0-aaa1-503df99cf13f this
	Member [127.0.0.1]:5704 - a1246091-68ac-407d-951b-8e07dfdb427b
]

12:03:28,286  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5702 [dev] [4.1.2-SNAPSHOT] Shutting down node engine...
12:03:28,287  WARN |clusterShouldBeEventuallySafe_withPartitionInitialization| - [InternalPartitionService] hz.relaxed_villani.generic-operation.thread-4 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Following unknown addresses are found in partition table sent from master[[127.0.0.1]:5701]. (Probably they have recently joined or left the cluster.) {
	[127.0.0.1]:5702 - c171e63c-d0b4-4794-ac89-682e958bd307
}
12:03:28,286  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [TransactionManagerService] hz.sad_villani.cached.thread-2 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5702, UUID: c171e63c-d0b4-4794-ac89-682e958bd307
12:03:28,288  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [ClusterService] hz.relaxed_villani.generic-operation.thread-6 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] 

Members {size:2, ver:6} [
	Member [127.0.0.1]:5701 - ea98e464-aebd-44d0-aaa1-503df99cf13f
	Member [127.0.0.1]:5704 - a1246091-68ac-407d-951b-8e07dfdb427b this
]

12:03:28,288  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [TransactionManagerService] hz.relaxed_villani.cached.thread-11 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5702, UUID: c171e63c-d0b4-4794-ac89-682e958bd307
12:03:28,289  WARN |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MigrationManager] hz.sad_villani.migration - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Cannot commit promotions. Destination [127.0.0.1]:5702 - c171e63c-d0b4-4794-ac89-682e958bd307 is not a member anymore
12:03:28,289  WARN |clusterShouldBeEventuallySafe_withPartitionInitialization| - [InternalPartitionService] hz.relaxed_villani.priority-generic-operation.thread-0 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Following unknown addresses are found in partition table sent from master[[127.0.0.1]:5701]. (Probably they have recently joined or left the cluster.) {
	[127.0.0.1]:5702 - c171e63c-d0b4-4794-ac89-682e958bd307
}
12:03:28,289  WARN |clusterShouldBeEventuallySafe_withPartitionInitialization| - [InternalPartitionService] hz.relaxed_villani.generic-operation.thread-11 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Following unknown addresses are found in partition table sent from master[[127.0.0.1]:5701]. (Probably they have recently joined or left the cluster.) {
	[127.0.0.1]:5702 - c171e63c-d0b4-4794-ac89-682e958bd307
}
12:03:28,291  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [NodeExtension] Thread-141 - [127.0.0.1]:5702 [dev] [4.1.2-SNAPSHOT] Destroying node NodeExtension.
12:03:28,291  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5702 [dev] [4.1.2-SNAPSHOT] Hazelcast Shutdown is completed in 6 ms.
12:03:28,291  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [LifecycleService] Thread-141 - [127.0.0.1]:5702 [dev] [4.1.2-SNAPSHOT] [127.0.0.1]:5702 is SHUTDOWN
12:03:28,292  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [LifecycleService] Thread-141 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] [127.0.0.1]:5704 is SHUTTING_DOWN
12:03:28,292  WARN |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Terminating forcefully...
12:03:28,292  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Shutting down connection manager...
12:03:28,292  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MockServer] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5704, connection: MockConnection{localEndpoint=[127.0.0.1]:5701, remoteEndpoint=[127.0.0.1]:5704, alive=false}
12:03:28,292  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MockServer] Thread-141 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Removed connection to endpoint: [127.0.0.1]:5701, connection: MockConnection{localEndpoint=[127.0.0.1]:5704, remoteEndpoint=[127.0.0.1]:5701, alive=false}
12:03:28,292  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MembershipManager] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Removing Member [127.0.0.1]:5704 - a1246091-68ac-407d-951b-8e07dfdb427b
12:03:28,292  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [ClusterService] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] 

Members {size:1, ver:7} [
	Member [127.0.0.1]:5701 - ea98e464-aebd-44d0-aaa1-503df99cf13f this
]

12:03:28,292  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [TransactionManagerService] hz.sad_villani.cached.thread-6 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Committing/rolling-back live transactions of [127.0.0.1]:5704, UUID: a1246091-68ac-407d-951b-8e07dfdb427b
12:03:28,292  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Shutting down node engine...
12:03:28,294  WARN |clusterShouldBeEventuallySafe_withPartitionInitialization| - [MigrationManager] hz.sad_villani.migration - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Cannot commit promotions. Destination [127.0.0.1]:5704 - a1246091-68ac-407d-951b-8e07dfdb427b is not a member anymore
12:03:28,298  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [NodeExtension] Thread-141 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Destroying node NodeExtension.
12:03:28,298  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] Hazelcast Shutdown is completed in 6 ms.
12:03:28,298  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [LifecycleService] Thread-141 - [127.0.0.1]:5704 [dev] [4.1.2-SNAPSHOT] [127.0.0.1]:5704 is SHUTDOWN
12:03:28,298  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [LifecycleService] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] [127.0.0.1]:5701 is SHUTTING_DOWN
12:03:28,298  WARN |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Terminating forcefully...
12:03:28,298  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Shutting down connection manager...
12:03:28,298  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Shutting down node engine...
12:03:28,304  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [NodeExtension] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Destroying node NodeExtension.
12:03:28,304  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [Node] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] Hazelcast Shutdown is completed in 6 ms.
12:03:28,304  INFO |clusterShouldBeEventuallySafe_withPartitionInitialization| - [LifecycleService] Thread-141 - [127.0.0.1]:5701 [dev] [4.1.2-SNAPSHOT] [127.0.0.1]:5701 is SHUTDOWN
BuildInfo right after clusterShouldBeEventuallySafe_withPartitionInitialization(com.hazelcast.internal.partition.PartitionServiceSafetyCheckTest): BuildInfo{version='4.1.2-SNAPSHOT', build='20210224', buildNumber=20210224, revision=d54afe8, enterprise=false, serializationVersion=1}
Hiccups measured while running test 'clusterShouldBeEventuallySafe_withPartitionInitialization(com.hazelcast.internal.partition.PartitionServiceSafetyCheckTest):'
12:01:25, accumulated pauses: 1888 ms, max pause: 316 ms, pauses over 1000 ms: 0
12:01:30, accumulated pauses: 2060 ms, max pause: 225 ms, pauses over 1000 ms: 0
12:01:35, accumulated pauses: 2305 ms, max pause: 198 ms, pauses over 1000 ms: 0
12:01:40, accumulated pauses: 1439 ms, max pause: 100 ms, pauses over 1000 ms: 0
12:01:45, accumulated pauses: 1313 ms, max pause: 51 ms, pauses over 1000 ms: 0
12:01:50, accumulated pauses: 892 ms, max pause: 53 ms, pauses over 1000 ms: 0
12:01:55, accumulated pauses: 575 ms, max pause: 39 ms, pauses over 1000 ms: 0
12:02:00, accumulated pauses: 117 ms, max pause: 19 ms, pauses over 1000 ms: 0
12:02:05, accumulated pauses: 284 ms, max pause: 105 ms, pauses over 1000 ms: 0
12:02:10, accumulated pauses: 186 ms, max pause: 27 ms, pauses over 1000 ms: 0
12:02:15, accumulated pauses: 99 ms, max pause: 16 ms, pauses over 1000 ms: 0
12:02:20, accumulated pauses: 52 ms, max pause: 7 ms, pauses over 1000 ms: 0
12:02:25, accumulated pauses: 58 ms, max pause: 6 ms, pauses over 1000 ms: 0
12:02:30, accumulated pauses: 306 ms, max pause: 33 ms, pauses over 1000 ms: 0
12:02:35, accumulated pauses: 165 ms, max pause: 26 ms, pauses over 1000 ms: 0
12:02:40, accumulated pauses: 288 ms, max pause: 39 ms, pauses over 1000 ms: 0
12:02:45, accumulated pauses: 75 ms, max pause: 9 ms, pauses over 1000 ms: 0
12:02:50, accumulated pauses: 88 ms, max pause: 24 ms, pauses over 1000 ms: 0
12:02:55, accumulated pauses: 76 ms, max pause: 17 ms, pauses over 1000 ms: 0
12:03:00, accumulated pauses: 89 ms, max pause: 13 ms, pauses over 1000 ms: 0
12:03:05, accumulated pauses: 37 ms, max pause: 3 ms, pauses over 1000 ms: 0
12:03:10, accumulated pauses: 46 ms, max pause: 11 ms, pauses over 1000 ms: 0
12:03:15, accumulated pauses: 266 ms, max pause: 37 ms, pauses over 1000 ms: 0
12:03:20, accumulated pauses: 282 ms, max pause: 39 ms, pauses over 1000 ms: 0
12:03:25, accumulated pauses: 219 ms, max pause: 168 ms, pauses over 1000 ms: 0
```

--- a/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/AbstractPartitionPrimaryReplicaAntiEntropyTask.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/partition/impl/AbstractPartitionPrimaryReplicaAntiEntropyTask.java
@@ -26,6 +26,7 @@ import com.hazelcast.internal.services.ServiceNamespace;
 import com.hazelcast.logging.ILogger;
 import com.hazelcast.spi.impl.NodeEngineImpl;
 import com.hazelcast.spi.impl.PartitionSpecificRunnable;
+import com.hazelcast.spi.impl.executionservice.ExecutionService;
 import com.hazelcast.spi.impl.operationservice.Operation;
 import com.hazelcast.spi.impl.operationservice.OperationService;
 import com.hazelcast.spi.impl.operationservice.UrgentSystemOperation;
@@ -34,6 +35,7 @@ import java.util.Collection;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Map;
+import java.util.concurrent.ExecutorService;
 import java.util.function.BiConsumer;
 
 import static com.hazelcast.internal.partition.IPartitionService.SERVICE_NAME;
@@ -119,11 +121,13 @@ public abstract class AbstractPartitionPrimaryReplicaAntiEntropyTask
 
         OperationService operationService = nodeEngine.getOperationService();
         if (hasCallback) {
+            ExecutorService asyncExecutor =
+                    nodeEngine.getExecutionService().getExecutor(ExecutionService.ASYNC_EXECUTOR);
             operationService.createInvocationBuilder(SERVICE_NAME, op, target.address())
                     .setTryCount(OPERATION_TRY_COUNT)
                     .setTryPauseMillis(OPERATION_TRY_PAUSE_MILLIS)
                     .invoke()
-                    .whenCompleteAsync(callback);
+                    .whenCompleteAsync(callback, asyncExecutor);
         } else {
             operationService.send(op, target.address());
         }

--- a/hazelcast/src/test/java/com/hazelcast/internal/partition/PartitionServiceSafetyCheckTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/internal/partition/PartitionServiceSafetyCheckTest.java
@@ -23,12 +23,16 @@ import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.annotation.ParallelJVMTest;
 import com.hazelcast.test.annotation.QuickTest;
+import com.hazelcast.test.annotation.SlowTest;
 import org.junit.Before;
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
 import java.util.Collection;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ForkJoinPool;
 import java.util.concurrent.TimeUnit;
 
 import static com.hazelcast.internal.partition.AntiEntropyCorrectnessTest.setBackupPacketDropFilter;
@@ -160,6 +164,52 @@ public class PartitionServiceSafetyCheckTest extends PartitionCorrectnessTestSup
                 }
             }
         });
+    }
+
+    @Category(SlowTest.class)
+    @Test(timeout = 60000)
+    public void clusterSafe_completes_whenInvokedFromCommonPool() throws InterruptedException {
+        Config config = getConfig(true, true);
+
+        HazelcastInstance hz = factory.newHazelcastInstance(config);
+        startNodes(config, nodeCount);
+
+        final Collection<HazelcastInstance> instances = factory.getAllHazelcastInstances();
+
+        // ensure we start enough threads to block all threads of FJP#commonPool
+        int threadCount = ForkJoinPool.commonPool().getParallelism() + 2;
+        spawn(() -> {
+            // trigger partition assignment while many threads are querying cluster safety
+            sleepSeconds(5);
+            warmUpPartitions(instances);
+            fillData(hz);
+        });
+        ExecutorService executorService = ForkJoinPool.commonPool();
+        // no assertions are actually executed. We only care that isClusterSafe
+        // does not hang indefinitely
+        assertTrueAllTheTime(() -> {
+            CountDownLatch startQueryClusterSafe = new CountDownLatch(1);
+            CountDownLatch completed = new CountDownLatch(threadCount);
+            for (int i = 0; i < threadCount; i++) {
+                executorService.submit(() -> {
+                            try {
+                                startQueryClusterSafe.await();
+                                for (HazelcastInstance instance : instances) {
+                                    PartitionService ps = instance.getPartitionService();
+                                    // just invoke isClusterSafe, no need to assert result
+                                    // we only care about the call eventually completing
+                                    ps.isClusterSafe();
+                                }
+                            } catch (InterruptedException e) {
+                                e.printStackTrace();
+                            } finally {
+                                completed.countDown();
+                            }
+                        });
+            }
+            startQueryClusterSafe.countDown();
+            completed.await();
+        }, 30);
     }
 
     @Test

